File: Baekjoon/Gold/2293.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [n, k] = input[0].split(" ").map(Number);
const v = input.slice(1).map(Number).sort();

let cache = new Array(k + 1).fill(0);
cache[0] = 1;

function cal(v) {
  let i = 0;
  while (i + v <= k) {
    cache[i + v] += cache[i];
    i++;
  }
}

v.map((value) => cal(value));

console.log(cache[k]);

File: Baekjoon/Gold/2293.py
nk = input().split(' ')
n = int(nk[0])
k = int(nk[1])
v = []
for i in range(n):
    v.append(int(input()))

v.sort()
cache = [0] * (k+1)
cache[0] = 1

def cal(v):
    i = 0
    while (i + v <= k):
        cache[i+v] += cache[i]
        i += 1

for value in v:
    cal(value)

print(cache[k])
